### DISCUSSION TITLE
what's an appropriate design pattern for combining embind + pthreads + async?

### DISCUSSION THREAD

**Discussion by dr-matt** [emscripten-core/emscripten#21422] [Q&A]
Hi, I am doing real-time data processing via wasm / c++ and looking to improve efficiency. What I'd like to have happen is this:

1. JS code instantiates stateful c++ class (wrapped using embind) at startup.
2. When a given data packet is received, JS passes it to the class instance via an embound class method.
3. That method returns a `Promise` immediately so as not to block the main thread.
4. Behind the scenes, the invoked c++ method spawns a thread to do the actual work (so that multiple sequential packets can be processed simultaneously), and when that work is done, it resolves the `Promise` object that was returned in step 3.

So in JS land, something like:
```javascript
const processor = new MyModule.Processor();  // create instance of stateful class

while (getNextPacket()) {
  if (processor.isBusy()) continue;
  processor.processData(newData)  // invoke class method that returns a promise
  .then((result) => {
    doSomething(result);
  });
}

processor.delete();
```

In c++:
```c++
emscripten::val Processor::process_data(const emscripten::val& input_data) {
  auto future = std::async(std::launch::async, do_work, input_data);  // spawn thread to do work
  // TODO: return promise linked to the spawned task
}

emscripten::val do_work(const emscripten::val& input_data) {
  // ... heavy computation here
  // TODO: when done, somehow resolve the promise returned above
}

EMSCIRPTEN_BINDINGS(my_module) {
  emscripten::class_<Processor>("Processor")
    .constructor<>()
    .function("processData", &Processor::process_data)
  ;
}
```

Basically looking for pointers on how to accomplish the above in a correct and hopefully simple way, as there are interactions between JS async, pthreads, and embind class bindings.

Using the c++ `future` as written will block the main thread - what's the right pattern to avoid this? I'd like to keep all the thread operations in c++, rather than say creating a separate JS worker and manually passing messages back and forth; and since there is no `main()` here, the `-sPROXY_TO_PTHREAD` option isn't viable. Also, for performance reasons I'd like to avoid using asyncify if possible, and for simplicity I'd like to use the c++ standard library (`std::thread`, `std::future`, `std::async`, etc) rather than lower-level pthread or emscripten calls.

How can a JS `Promise` object be returned from `process_data()` ? I've had a look at @RReverser 's c++20 coroutine integration as described [here](https://github.com/emscripten-core/emscripten/pull/20420), #20420 which is really interesting and possibly relevant for this use case, but I don't understand it well enough to know how/whether it can work in conjunction with threads.

Lastly - how can that `Promise` be resolved after the work is complete, so that the JS promise chain can continue execution?

Any help is appreciated. Thanks!

**Comment by sbc100** [comment]
You probably want to take a look the proxying API: https://emscripten.org/docs/api_reference/proxying.h.html

And also the C/C++ interface to JS promises: https://github.com/emscripten-core/emscripten/blob/main/system/include/emscripten/promise.h

In particular there is an API in promise.h called `emscripten_proxy_promise` which can be used to create a promise that resolves once the proxied work is complete.   The return value from `emscripten_proxy_promise` is an integer handle but in you JS code you can call `getPromise` to get a the actual JS handle: https://github.com/emscripten-core/emscripten/blob/b5b7fedda835bdf8f172a700726109a4a3899909/src/library_promise.js#L12

@tlively am I missing anything here?

**Reply by dr-matt** [reply]
@sbc100 thank you for the pointers! I will study `proxying.h` and `promise.h` in light of your suggestions. On first read, these look like fairly low-level APIs; do you know if the promise can somehow be resolved with `emscripten::val` as the result using these APIs? And can embind still be used to simply wrap the c++ method so that it returns a proper `Promise` to JS? I'll tinker with using `EM_JS` to call `getPromise()` as a start.

I'll also have a look at the corresponding unit tests for these APIs to get an idea of how it all fits together, but any other examples would be fantastic.

**Reply by dr-matt** [reply]
Where and how should `getPromise` be called? Is that just an emscripten-internal function or is it somehow also accessible to the caller JS (or preferably, to `EM_JS` in the c++ code)?

**Reply by sbc100** [reply]
`getPromise` is a JS library function.   JS library functions can be called from any JS code, including `EM_JS` or `EM_ASM` blocks.  However in order to have them included you need to tell the compiler to include them either via the `DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` command line setting of the `EM_JS_DEPS` macro, or via a dependency of some of other JS library function.

**Reply by sbc100** [reply]
In terms of higher level embind-style promise integration, I think @RReverser worked on some stuff like that, although I'm not sure it integrates seamlessly with proxying?

**Reply by dr-matt** [reply]
@sbc100 thank you, the `EM_JS_DEPS` is what I was missing and now I'm able to wrap the promise in a `val` object and embind it for return to JS.

I did have a look at @RReverser 's great work on c++ coroutines and tried it out. It worked very easily for single-threaded examples, but I couldn't figure out how to get it working across threads. I did not try combining it with proxying, but admittedly don't know enough about either API to integrate them properly.

In the meantime, with your help, I was able to make a lot more progress using the lower-level `emscripten_proxy_promise` API. I can now add tasks to the work queue on various spawned threads and return a valid `Promise` back to JS, which resolves when the task completes. The last major missing piece is how to resolve the promise with the desired JS `val` object from c++. How does the proxy worker thread pass a value back to the promise handle once it's resolved, and then properly interpret that value for JS, since it looks to just be a generic `void*` ?

**Reply by tlively** [reply]
Since there is no way to share JS objects between threads, there's no way to share `val` between threads, either. I would do all your multithreading and proxying with C++ objects, then convert to/from val as necessary on each thread individually.

To pass C++ objects via that `void*`, you would have to allocate them on the heap, cast the pointer to `void*`, then cast back to a pointer of the proper type and where it is received. Don't forget to `free` or `delete` the object when you are done with it.

**Reply by dr-matt** [reply]
@tlively that makes sense, thank you. `val` not being shareable between threads explains why previous attempts at this did not work the way I thought they should. A few more questions:

The callback given to `emscripten_proxy_promise` only accepts the `void*` user data and returns `void`. It seems odd that this method doesn't accept an `em_promise_callback_t` instead, which would allow resolution of the promise (`FULFILL`, `REJECT`) as well as returning a resolved value. How is this callback supposed to supply any result output?

To get data back to the main thread's returned `Promise`, I'm using a call to `emscripten_promise_then` chained after `emscripten_proxy_promise`. The former does take an `em_promise_callback_t` as an input, unlike the latter. That callback has three inputs: `void** result`, `void* data`, and `void* value`. The `value`, which is the result of the upstream promise (in this case the proxied one), is always `NULL` and the promise is always `FULFILLED`. Related to the first question above, can this behavior be changed directly, or is it expected?

Lastly - I can return a pointer by setting `*result = result_pointer`, and I do see `result_pointer` as an integer on the JS side in the resolved promise. How can I return an actual value here instead, like a `val` object or JS object, or even a JS primitive?

For reference and review, here's what I have so far:
```c++

// helper to get an actual promise object from an em_promise_t
EM_JS(emscripten::EM_VAL, get_js_promise, (int handle), {
  return Emval.toHandle(getPromise(handle));
});

class MyClass {
  ...
  // this class method is exposed to js via embind
  emscripten::val test_proxy_promise(const emscripten::val& i_string) {
    // define data type to hold input and output of proxied function
    struct Data {
      std::string input_string;
      std::string output_string;
    };
    auto input_data = new Data();
    input_data->input_string = i_string.as<std::string>();

    // start up separate thread to do the work
    std::thread thread([this]() {
      while (!done) {
        emscripten_proxy_execute_queue(m_proxying_queue);
        std::this_thread::yield();
      }
    });
    const auto thread_handle = thread.native_handle();
    thread.detach();

    // this is the proxied function
    auto do_work = [](void* i_data) {
      Data& data = *reinterpret_cast<Data*>(i_data);
      data.output_string = data.input_string + "_NEW";
    };

    auto promise = emscripten_proxy_promise(m_proxying_queue, thread_handle, do_work, input_data);

    // "then" callback to populate output result of resolved promise
    auto populate_callback = [](void** o_result, void* i_data, void* i_value) {
      Data& data = *reinterpret_cast<Data*>(i_data);
      emscripten::val* new_val = new emscripten::val(data.output_string);
      *o_result = new_val;
      delete &data;  // don't need this anymore
      return EM_PROMISE_FULFILL;
    };

    // dummy callback for upstream rejection
    auto error_callback = [](void** result, void* data, void* value) {
      return EM_PROMISE_REJECT;
    };

    // chain another promise onto the proxied one so we can get a result on the main thread
    auto then = emscripten_promise_then(promise, populate_callback, error_callback, input_data);

    // get promise object as a val
    auto promise_val = emscripten::val::take_ownership(get_js_promise(reinterpret_cast<int>(then)));

    // clean up promises
    emscripten_promise_destroy(promise);
    emscripten_promise_destroy(then);

    // return val to JS via embind
    return promise_val;
  }
  ...
}
```

**Reply by tlively** [reply]
You've got the right idea where you use the same context object (`input_data` in this case) as the `arg` to `emscripten_proxy_promise` and the `data` for `emscripten_promise_then`. Any value you want to flow from the proxied work to a chained promise should be written into that context object, just like you are doing.

To logically return something other than an integer using the low-level APIs, you would have to use an indirection. For example, you could return a pointer to a heap-allocated `val` or you could return an index into an array of JS objects.

I believe the promise `val` supports being resolved with arbitrary other `val`s, though.

**Reply by dr-matt** [reply]
@tlively thanks! Glad to hear this is on the right track. Everything seems to be working as desired now, except that I'm still unclear as to how to get a non-integer value (in this case `new_val`, which is a pointer to a heap-allocated `val` as you mentioned) back into `promise_val`, or reconstitute the integer pointer in `promise_val` to a JS object.

It's also unclear how to manage the lifecycle of the heap-allocated return value (e.g., `new_val`) without leaking memory - how and when can this value be deleted?

I really appreciate your help.

**Reply by tlively** [reply]
Could you do something like this to turn `promise_val` into another `val` promise that returns the dereferenced `new_val`?

```
emscripten::val deref(emscripten::val ptrPromise) {
  auto* ptr = reinterpret_cast<emscripten::val*>(co_await ptrPromise);
  auto val = *ptr;
  delete ptr;
  co_return val;
}
```

Unfortunately it can be difficult to manage the lifetimes of all this heap-allocated data, but I think you have the right idea there, too, where you `delete` the heap allocations in callbacks after you are done using them.

**Reply by dr-matt** [reply]
@tlively nice idea! That extra `deref` wrapper does indeed work, and gives me back the correct value in JS. I'm going to run with this solution for now. Many thanks, and also to @sbc100 for suggesting the overall method.

Having read up and tinkered a bit with coroutines and `co_await` on `emscripten::val`, it does feel like one should be able to accomplish this thread proxying using only higher-level c++ mechanisms rather than the lower-level C calls. Does anyone (including @RReverser) have thoughts on whether this could work and if so, how it might be constructed?

**Reply by dr-matt** [reply]
The examples in `test_pthread_proxying.h` contain other calls that I'm not currently including, such as `emscripten_exit_with_live_runtime` and `emscripten_runtime_keepalive_push`/`pop`. Are any of those needed in the above code? Also, what is the use case for the `with_ctx` version of `emscripten_proxy_promise` vs. without the context?

**Reply by tlively** [reply]
Since your worker thread is just spinning in a loop, there's no need to use those extra calls to keep the runtime alive. Those are for when you allow the worker thread main function to return so the worker returns to the event loop, but you still want to keep its runtime alive so that it will execute proxied work that comes in. It actually might be worth switching to this pattern because it is more efficient than having the worker spin in a loop.

The `with_ctx` versions of the proxying functions are for when the proxied work involves asynchronous JS, so it actually requires multiple turns of the worker's event loop to be finished. It lets the worker signal explicitly when the work is done rather than automatically considering the work to be done when the proxied function returns.

**Reply by dr-matt** [reply]
@tlively I think I understand. So I probably don't need the `with_ctx` calls, but the code may gain some efficiency from using the `runtime` related calls. I'll study those more to see if I can figure out how they work and how they should be incorporated in place of the spin loop.

**Reply by dr-matt** [reply]
Just getting back to this now - the one example I see of this model is in `test_pthread_proxying.c:358`:
```c++
em_promise_result_t
test_promise_returner(void** result, void* data, void* value) {
  widget* w16 = data;

  em_promise_t promise =
    emscripten_proxy_promise(proxy_queue, returner, do_run_widget, w16);

  *result = emscripten_promise_then(promise, check_widget, explode, w16);
  emscripten_promise_destroy(promise);

  return EM_PROMISE_MATCH_RELEASE;
}
```
The `returner` thread calls `returner_main`, which looks like:
 ```c++
void* returner_main(void* queue) {
  emscripten_err("returner_main");
  emscripten_exit_with_live_runtime();
}
```

As far as I can tell, the `returner` thread doesn't use the `proxy_queue` passed into `emscripten_proxy_promise` in this case, unlike the `looper` thread in that same file which is what I based my example on.

When I try this model, putting custom code in place of the `emscripten_err` call in `returner_main`, the function is called, but never returns from `emscripten_exit_with_live_runtime()` and that thread just hangs. So I don't quite see how all this works together. Is the thread function supposed to exit, and behind the scenes the corresponding worker is kept around for later somehow? Am I missing a function call or piece of the required logic?

**Reply by tlively** [reply]
The neat thing with this pattern is that the worker thread doesn't need to do anything to explicitly execute the proxied work. It will automatically wake up and execute the work as long as it returns to the JS event loop. `emscripten_exit_with_live_runtime` makes the worker thread stay alive while it exits the thread main function and returns to the event loop. When it is done with the work, it will return to the event loop and will be ready to execute the next proxied work that comes in until you explicitly force the thread to exit.
